cpp/pagerank_module/algorithm/pagerank.hpp

```diff
@@ -0,0 +1,189 @@
+#pragma once
+
+#include <future>
+
+namespace pagerank_util {
+
+/// @tparam T Node ID data type.
+template <typename T>
+class AdjacencyList {
+ public:
+  AdjacencyList() = default;
+  explicit AdjacencyList(std::uint64_t node_count) : list_(node_count) {}
+
+  /// If adjacency list size is unknown at the construction time this method
+  /// can be used to reserve the required space. The method will also clear
+  /// underlying storage if it contains something.
+  void Init(std::uint64_t node_count) {
+    list_.clear();
+    list_.resize(node_count);
+  }
```

Comment:
Not sure, but you could use `std::optional<AdjacencyList>` to eliminate the use case for a function like this.

Reply:
Can you elaborate on how could `optional` solve the issue of having such a method? Isn't the case when you need such a method the one where you want to use the same object (`AdjacencyList` here), but also want to start off with clean containers?

Reply:
You create
```
std::optional<AdjacencyList> adjecancy_list;
```

you call `adjecancy_list.emplace()` to create a new AdjecancyList (all containers clean).
You can call `adjecancy_list.emplace()` again to erase the old one and create a new AdjecancyList, again with clean containers.

Reply:
But if you want to keep the same API, I would rename it to something like `Reinitialize`.

Reply:
I wasn't aware of that functionality with `std::optinal`, definitely, I'll have it in mind, although it would complicate things here, where basically we don't have a reason for reinitialization.

Reply:
Yep, you don't so that's why I would remove it entirely.